Add hook specification to add new icons and stylesheets

lukasz-migas opened the pull request:
This PR adds new (experimental) hook specification `napari_experimental_provide_theme` to provide a way to add new themes, new icons (or replace existing ones) and add extra stylesheets. Looking through the codebase I could not find an easy way to modify stylesheets or easily add extra icons (I need this for the 1d plugin I am working on to add new layers) so I thought adding plugin specification might be one way to do it. 

The new hook simply returns a list of stylesheet paths, list of svg icon paths and dictionary of theme(s) which are then added in appropriate places and used to generate the `_qt_resources_...py` file.

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How has this been tested?
Some tests are failing on my machine but I think it's unrelated to my changes.

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

Somewhat addresses #936.

lukasz-migas committed:
Add hook specification to add new themes, icons and stylesheets

- added new hook specification to allow addition of new themes, icons and stylesheets to napari
- minor improvements to how qt resources are loaded

lukasz-migas committed:
Update docstrings

- updated docstrings

tlambert03 commented on napari/plugins/hook_specifications.py:
```diff
@@ -496,3 +496,33 @@ def napari_experimental_provide_dock_widget() -> Union[
+    -------
+    qss_files : List[str]
+        A list of Qt stylesheets with the file extension .qss. Napari provides
+        several default stylesheets with names `00_base.qss`, `01_buttons.qss` etc
```
this makes me wonder if we should name our files `100_base.qss`, `200_buttons.qss`, etc...

tlambert03 commented on napari/plugins/hook_specifications.py:
```diff
@@ -496,3 +496,33 @@ def napari_experimental_provide_dock_widget() -> Union[
+
+    Returns
+    -------
+    qss_files : List[str]
```
if we stick with this, we should provide an example of the theme name template expansion that we use in our qss ... like `background-color: {{ background }};`

lukasz-migas replied:
Sure, that's a good idea. I will expand the docs in the function to make it clearer. I think I might go with @sofroniewn idea of splitting it into two separate functions with one available to provide themes (just dict of colors) and another for stylesheets and icons.

tlambert03 commented on napari/plugins/theme.py:
```diff
@@ -0,0 +1,155 @@
+logger = getLogger(__name__)
+
+
+def get_stylesheet_from_plugins(
```
It looks like you loosely patterned `get_stylesheet_from_plugins` after our reader plugin "call hook"... which is probably not the best one to use here.  That one loops through all of the plugins every time, but for this one, it shouldn't be necessary to perform this functionality any more than once per package.  For that, the plugin engine provides the "call_historic" approach, which essentially let's us "call" the hook a single time (at load) and then when each plugin gets registered for the first time, a registration callback will be called for that specific plugin.  See, for example how we register the `dock_widgets`, starting [here](https://github.com/napari/napari/blob/ed6b8ef2e55471b42a3ea47875b7e1ef0f99ba80/napari/plugins/_plugin_manager.py#L419)

1. `discover_widgets` calls the actual hook with `self.hook.<specname>.call_historic()`.  this is in contrast to iterating through `get_hookimpls()` as we do with reader plugins.  The _default_ behavior of the plugin engine is actually to call all the hooks, so that whole looping we do in the reader plugin is a special/odd case.  (note, the call to `discover_widgets` actually happens only once our `qt_main_window` is instantiated for the first time).
2. Each time a plugin is registered, if it provides a `<specname>` implementation, the function provided to `call_historic` will be called (in the case of dock widgets, it's [`register_dock_widget`](https://github.com/napari/napari/blob/ed6b8ef2e55471b42a3ea47875b7e1ef0f99ba80/napari/plugins/_plugin_manager.py#L304)).  The correlate here would be a function that is responsible for performing the qss & theme dict aggregation, and compiling the resources.  (It would also do all the return value validation).

This _would_ essentially mean that a lot of the resource compilation would need to be refactored a little bit.  That is, rather than gathering _all_ of the icons each time and recompiling them anytime anything is different, the callback provided to `call_historic` would ideally be able to compile just the new svgs for each plugin.  (I'm not sure if that would come with much of a performance penalty... but it would make it cleaner to register each plugin individually).

lukasz-migas replied:
Thanks for the pointers and ideas. I think with small number of theme/stylesheet the current approach would lead to significant performance loss but it can be cleaner. I will look into changing the behaviour to discover themes etc once at the startup rather than recompile all of the resources each time a new plugin is installed.

tlambert03 reviewed (commented):
This is a great start @lukasz-migas!  
The hook specification as a list of QSS, icons, and theme dicts certainly mimics what we've done internally, though I do wonder if they should perhaps give us a single dict rather than a tuple?  it might be easier to modify in the future.

```python
{
    'qss': [],
    'icons': [],
    'palletes': {
        'super_dark': {
            "folder": "super_dark",
            "background": "rgb(12, 12, 12)",
        }
    }
}
```

more work is needed to integrate with preferences.  currently, the new theme won't show up in the preferences window... and if you set it programatically... then the preferences window fails to open.  (this may well be due to some poor assumptions in the preferences window... but needs to be addressed).

also, it would be nice if a new theme could be registered after the original bootstrapping is done (i.e. if a plugin was installed during runtime, it could be available immediately).  For that, I think the `call_historic` approach that I mention below might be better.

Lastly, you've done all this in a very small amount of code changes, which is nice.  But i wonder if perhaps we should be re-factoring napari's internals more to use this spec as we've done with other hook specs.  That is, rather than putting plugins into napari's internal `STYLES` and `ICONS` and stuff, and then calling `_register_napari_resources` ... we might want to remove the `_register_napari_resources` altogether, and use the lower level `compile_qt_svgs` both for plugins  (That was what that function was originally conceived for, and an example usage is [here](https://github.com/napari/napari/blob/master/examples/dev/custom_icon.py)) _and_ for napari as well, with a "builtin" plugin implementation.

I'm not sure everything here was coherent :) ... so if you'd like to jump on a zoom or zulip to chat about this let me know.

sofroniewn commented on napari/plugins/hook_specifications.py:
```diff
@@ -496,3 +496,33 @@ def napari_experimental_provide_dock_widget() -> Union[
+    This hook specification is marked as experimental as the API or how the returned
+    value is handled may change here more frequently than the rest of the codebase.
+
+    Returns
```
I might want to reorder this tuple here so that `color_dict` comes first, then `qss_files`, and then `svg_paths` as I see `color_dict` as the simplest most widely accessible option. It's not to much to ask folks who don't want to provide the other things to put in empty lists, but we could also think about just allowing the `Dict[str, Dict[str, str]]` for people who only want to return a `color_dict`. Not sure I need that, but putting it out there

sofroniewn replied:
Actually thinking further I wonder if we want to have two hookspecs here. One for providing the color dictionary and one for providing the custom qss/ icon svgs.

I have two reasons for this. One is that that the color dict is independent of Qt so could be very stable (and be reused if we ever did another front end). And two at any moment in time you only want to have one theme/ color_dict be active, and these choices could populate the settings and preferences (cc @goanpeca @ppwadhwa according to Lia's design). These are then thought of in an "exclusive" manner, at any one time you only use the output of one plugin or our builtin.

In contrast, the additional qss sheets and icons are something that should be thought as additive, i.e. if you have 5 plugins the offering from all 5 should now be available, though one might need to think about conflicts here ...

Not sure if this is really the right approach either, but curious what others think too.

lukasz-migas replied:
As I mentioned above, I think I will split it into two separate hooks since it's probably an overkill to always have to provide the icons/stylesheets if only color scheme is being changed. 

> In contrast, the additional qss sheets and icons are something that should be thought as additive, i.e. if you have 5 plugins the offering from all 5 should now be available, though one might need to think about conflicts here ...

Indeed, this is my concern that if multiple plugins are providing icons/stylesheets/themes then it could become a bit of a mess. Perhaps they could be ordered like the `reader/writer/etc` plugins but that might not be that elegant.

sofroniewn replied:
Yeah, one concrete use case for adding icons is to let plugins add their own icon/ and "about" text description that we generate a widget from. This might be a new hookspec as we could generate the widget for them and show it at various places in the app, but wanted to mention it. I know @tlambert03 and I have discussed it in other places (cc @neuromusic)

sofroniewn reviewed (commented):
Love this as well, only had time for a quick look this evening, and left a couple initial reactions, but we should be able to work this all out fairly easily!!

lukasz-migas committed:
Merge branch 'napari:master' into theme-hook

lukasz-migas committed:
WIP: added evented dict container, split hook spec into three hooks

- split the hook implementation into three separate hooks (qss, icons and theme)
- added new container class: `EventedDict` and `TypedMutableMapping` so we can migrate themes to an evented class
- refactored how theme plugin data is read

sofroniewn commented on napari/utils/events/containers/_evented_dict.py:
```diff
@@ -0,0 +1,107 @@
+from ._dict import _K, _T, TypedMutableMapping
+
+
+class EventedDict(TypedMutableMapping[_K, _T]):
```
Wow love this @lukasz-migas! Thanks for adding

sofroniewn commented on napari/plugins/_plugin_manager.py:
```diff
@@ -292,6 +293,81 @@ def available_samples(self) -> Tuple[Tuple[str, str], ...]:
+
+        Each key in `data` is a `theme_name` and the value is dictionary of values.
+        """
+        keys = [
```
I wonder if we should define this list of keys somewhere in our `utils/theme.py` file and import them here. Seems less likely for something to get out of sync if we were to ever change any of these

lukasz-migas replied:
That makes sense. I've changed the `_themes` dict to use the `EventedDict` and also perform key validation whenever new theme is added. Haven't fully tested it yet so it might change.

sofroniewn commented on examples/theme_plugin.py:
```diff
@@ -0,0 +1,33 @@
+"""Very simple theme plugin example providing simple color scheme changes."""
```
i wonder if we should put this in a new folder `examples/plugins`? As it's a little more for reference than for running, as if i run example nothing happens as no viewer is launched. Or maybe there's a way to launch viewer and show plugin being used in same script too?

lukasz-migas replied:
Since I am still making changes, I haven't worked out a good example yet. I think its a good idea to have sub-category of examples specifically for plugins. Once I iron out the kinks I will add one in.

sofroniewn reviewed (commented):
This is looking great! Thanks for adding the evented dict too!! I just left a couple minor comments. Let us know if this is ready for full review

lukasz-migas committed:
Theme dict is now evented, immediate updates to stylesheet

- theme dict is now an evented model
- changes to any theme data will result in immediate update of the stylesheet
- minor bug fix to the evented dict not propagating child events

lukasz-migas committed:
Merge branch 'napari:master' into theme-hook

lukasz-migas committed:
Themes now reflect state of plugin, preferences shows all available themes

- themes are now dynamically updated whenever a plugin is enabled/disabled/removed
- theme dropdown is now updated whenever new theme is added/removed
- plugin manager now displays more useful information if theme data failed validation

lukasz-migas committed:
Merge branch 'theme-hook' of https://github.com/lukasz-migas/napari into theme-hook

lukasz-migas committed:
Icon colors can be changed dynamically

- fixed issue in settings tests (the test should have always failed since it was passing wrong data to the `register_theme` function)
- found a workaround for dynamically updating icon colors without causing crash...

tlambert03 commented on napari/utils/theme.py:
```diff
@@ -29,38 +46,65 @@ def __getattr__(attr):
-        'syntax_style': 'default',
-        'console': 'rgb(255, 255, 255)',
-        'canvas': 'white',
+class ThemesDict(EventedDict):
```
The `EventedDict` is fantastic, and it'd be great if you could break it out into it's own PR so we can fill out our evented containers.

I'm certain that class will be used.  As mentioned in the meeting though, I wonder if we might be better served by making our `Theme` objects subclass from pydantic.  It even has a lovely `Color` field that we can use:

```python
from napari.utils.events import EventedModel
from pydantic.color import Color

# not sure how many of these we want to provide by default?
class Theme(EventedModel):
    name: str
    syntax_style: str = 'native'
    background: Color = 'rgb(38, 41, 48)'
    foreground: Color = 'rgb(65, 72, 81)'
    primary: Color = 'rgb(90, 98, 108)'
    secondary: Color = 'rgb(134, 142, 147)'
    highlight: Color = 'rgb(106, 115, 128)'
    text: Color = 'rgb(240, 241, 242)'
    icon: Color = 'rgb(209, 210, 212)'
    warning: Color = 'rgb(153, 18, 31)'
    current: Color = 'rgb(0, 122, 204)'
    console: Color = 'rgb(0, 0, 0)'
    canvas: Color = 'black'
```

```py
In [27]: t = Theme(name='whatever')

In [28]: t.events.foreground.connect(lambda e: print(e.value))

In [29]: t.foreground = 'azure'
azure

In [30]: t.foreground
Out[30]: Color('azure', rgb=(240, 255, 255))

In [31]: t.foreground.as_rgb()
Out[31]: 'rgb(240, 255, 255)'
```

what do you think?

lukasz-migas replied:
Thanks! I will make a separate PR for the `EventedDict` and update the underlying theme model. I was planning on using custom color converter (`transform_color` -> `rgb(X, Y, Z)`) but I think the pydantic's `Color` will work nicely! 😄

tlambert03 replied:
yeah, I'd love to see us use pydantic's `Color` field more liberally throughout the codebase.  In many cases, it could pretty much replace our [`standardize_color`](https://github.com/napari/napari/blob/master/napari/utils/colormaps/standardize_color.py) module ... and I suspect it's more tested by the community

sofroniewn replied:
> yeah, I'd love to see us use pydantic's Color field more liberally throughout the codebase. In many cases, it could pretty much replace our standardize_color module ... and I suspect it's more tested by the community

Yes - absolutely agree here. Maybe this is something @andy-sweet can look into too as he looks at some more of the color handling. It could remove a lot of our custom code

lukasz-migas committed:
Tidy up

- removed unused code

lukasz-migas committed:
[WIP] Add pydantic-based Theme, refactor theme

- added pydantic-based Theme class with better validation of values
- modified the way `get_theme` is retrieved. Currently it will trigger warnings whenever somebody calls it which might not be desired
- removed the `palettes` warning from utils/theme.py since it was meant to be removed in 0.4.6

lukasz-migas committed:
Update qt_main_window.py

- existing themes will now be properly connected at startup

lukasz-migas committed:
Added theme tests, theme improvements

- removed custom Themes model since its not necessary
- added a couple extra tests to handle new theme model (wip)
- improved docs regarding new theme hooks

lukasz-migas committed:
Temporarily remove evented dict

lukasz-migas committed:
Merge branch 'napari:master' into theme-hook

lukasz-migas committed:
Minor improvements

- changes to canvas/console and syntax style will now immediately update the ui
- minor changes to Theme model

lukasz-migas committed:
Disabled warning

lukasz-migas committed:
Update vispy_canvas.py

lukasz-migas committed:
Minor bug fixes

Fixes to `UserWarning` warnings when transforming color to 0-1 scale

lukasz-migas committed:
Undo formatting

lukasz-migas committed:
Removed `provide_theme` hook

- removed `napari_experimental_provide_theme` since its being handled by another PR ( #3034 )
- removed the Evented theme from this PR
- removed the plugin example since it couldn't be run anyway

lukasz-migas committed:
Minor

lukasz-migas committed:
Merge branch 'master' of https://github.com/napari/napari into napari-master

lukasz-migas committed:
Delete test_settings.py

lukasz-migas committed:
Merge branch 'napari-master' into theme-hook

lukasz-migas committed:
Update vispy_scale_bar_visual.py

lukasz-migas committed:
[wip] add namespace to icons

- icons now include plugin name so can be accessed via `plugin_name/icon_name.svg`
- renamed `plugin/theme` to `plugin/resources`

lukasz-migas committed:
Changed hooks to historic

- changed icons and qss hooks to be historic

lukasz-migas committed:
Merge branch 'napari:master' into theme-hook

lukasz-migas committed:
Added new icon namespace

- the `generate_colorized_svgs` method now accepts new argument `svg_stems` which acts in place of the auto-generated icon names. This enables addition of plugin-specific namespace
- removed unused code

lukasz-migas committed:
Update _icons.py

lukasz-migas committed:
Update _plugin_manager.py

lukasz-migas committed:
Bring up to date

lukasz-migas committed:
Added tests, fixed type annotations, general bug fixes

- update docs
- fix type annotations for `_provide_qss` and `_provide_icons` hook spec
- added examples to new hook specs
- improved type checking during registration
- added tests for both new hooks

lukasz-migas committed:
Move qss tests

- fix icons tests
- move plugin qss tests

lukasz-migas committed:
Stylesheets now use namespace

- moved stylesheets to own namespace including the pluginname
- updated docs
